module/VuFind/src/VuFind/Db/Table/Ratings.php

```diff
@@ -0,0 +1,181 @@
+ * @link     https://vufind.org Main Site
+ */
+class Ratings extends Gateway
+{
+    /**
+     * Constructor
+     *
+     * @param Adapter       $adapter Database adapter
+     * @param PluginManager $tm      Table manager
+     * @param array         $cfg     Laminas configuration
+     * @param RowGateway    $rowObj  Row prototype object (null for default)
+     * @param string        $table   Name of database table to interface with
+     */
+    public function __construct(
+        Adapter $adapter,
+        PluginManager $tm,
+        $cfg,
+        ?RowGateway $rowObj = null,
+        $table = 'ratings'
+    ) {
+        parent::__construct($adapter, $tm, $cfg, $rowObj, $table);
+    }
+
+    /**
+     * Get average rating and rating count associated with the specified resource.
+     *
+     * @param string $id     Record ID to look up
+     * @param string $source Source of record to look up
+     * @param ?int   $userId User ID, or null for all users
+     *
+     * @return array Array with keys count and rating (between 0 and 100)
+     */
+    public function getForResource(string $id, string $source, ?int $userId): array
+    {
+        $resourceTable = $this->getDbTable('Resource');
+        $resource = $resourceTable->findResource($id, $source, false);
+        if (empty($resource)) {
+            return [];
+        }
+
+        $callback = function ($select) use ($resource, $userId) {
+            $select->columns(
+                [
+                    'id' => new Expression(
+                        '1',
+                        [],
+                        [Expression::TYPE_IDENTIFIER]
+                    ),
+                    'count' => new Expression(
+                        'COUNT(?)',
+                        [Select::SQL_STAR],
+                        [Expression::TYPE_IDENTIFIER]
+                    ),
+                    'rating' => new Expression(
+                        'FLOOR(AVG(?))',
+                        ['rating'],
+                        [Expression::TYPE_IDENTIFIER]
+                    ),
+                ]
+            );
+            $select->where->equalTo('ratings.resource_id', $resource->id);
+            if (null !== $userId) {
+                $select->where->equalTo('ratings.user_id', $userId);
+            }
+        };
+
+        return $this->select($callback)->current()->toArray();
+    }
+
+    /**
+     * Delete a rating if the owner is logged in.  Returns true on success.
+     *
+     * @param int                 $id   ID of row to delete
+     * @param \VuFind\Db\Row\User $user Logged in user object
+     *
+     * @return bool
+     */
+    public function deleteIfOwnedByUser(int $id, \VuFind\Db\Row\User $user)
+    {
+        // User must have an ID:
+        if (!isset($user->id)) {
+            return false;
+        }
+
+        // Rating row must exist:
+        $matches = $this->select(['id' => $id]);
```

Comment:
Could you just include the user ID in the select here to save a step?

Reply:
Done. That was another copy-paste..